cmd/clusterctl/pkg/client/repository/repository_local_unix.go

```diff
@@ -0,0 +1,215 @@
+//
+// (1): {provider-name} must match the value returned by Provider.Name()
+// (2): {version} must exactly obey the syntax and semantics of
+// the "Semantic Versioning" specification (http://semver.org/).
+//
+// Concrete example:
+// /home/user/go/src/sigs.k8s.io/cluster-api-provider-aws/releases/v0.4.7/infrastructure-components.yaml
+// basepath: /home/user/go/src
+// owner : sigs.k8s.io
+// provider-name: cluster-api-provider-aws
+// version: v0.4.7
+// path/component.yaml: infrastructure-components.yaml
+//
+
+type localRepository struct {
+	providerConfig        config.Provider
+	configVariablesClient config.VariablesClient
+	basepath              string
+	owner                 string
+	providerName          string
+	defaultVersion        string
+	rootPath              string
+	componentsPath        string
+}
+
+var _ Repository = &localRepository{}
+
+// DefaultVersion returns defaultVersion field of localRepository struct
+func (r *localRepository) DefaultVersion() string {
+	return r.defaultVersion
+}
+
+// RootPath returns rootPath field of localRepository struct
+func (r *localRepository) RootPath() string {
+	return r.rootPath
+}
+
+// ComponentsPath returns componentsPath field of localRepository struct
+func (r *localRepository) ComponentsPath() string {
+	return r.componentsPath
+}
+
+// GetFile returns a file for a given provider version
+func (r *localRepository) GetFile(version, fileName string) ([]byte, error) {
+	var err error
+
+	if version == "latest" {
+		version, err = r.getLatestRelease()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get the latest release")
+		}
+	} else if version == "" {
+		version = r.defaultVersion
+	}
+
+	absolutePath := filepath.Join(r.basepath, r.owner, r.providerName, "releases", version, r.rootPath, fileName)
+
+	if f, err := os.Stat(absolutePath); err == nil {
+		if f.IsDir() {
+			return nil, errors.Errorf("invalid path: file %q is actually a directory %q", fileName, absolutePath)
+		}
+		content, err := ioutil.ReadFile(absolutePath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read files from local release %s", version)
+		}
+		return content, nil
+	}
+	return nil, errors.Errorf("failed to read files from local release %s, path %s", version, absolutePath)
+
+}
+
+// newLocalRepository returns a localRepository implementation
+func newLocalRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*localRepository, error) {
+
+	var err error
+
+	u, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+	absPath := u.Path
+	if u.RawPath != "" {
+		absPath = u.RawPath
+	}
+
+	if !filepath.IsAbs(absPath) {
+		return nil, errors.Errorf("invalid path: path %q must be an absolute path", providerConfig.URL())
+	}
+
+	parts := strings.Split(providerConfig.URL(), "/")
+	// find the first occurance of provider name and use it as a seperator between base and layout
```

Comment:
If we simplify the URL this part can be simplified as well and we can get rid of this search that can lead to problems if the provider name is repeated two or more times.

Reply:
Done.